// traveltime-api/target/error.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TravelTimeError = void 0;
class TravelTimeError extends Error {
    constructor({ http_status, error_code, description, documentation_link, additional_info, }) {
        super(description);
        this.name = 'TravelTimeError';
        this.http_status = http_status;
        this.error_code = error_code;
        this.description = description;
        this.documentation_link = documentation_link;
        this.additional_info = additional_info;
    }
    static isTravelTimeError(payload) {
        return !!payload.error_code && !!payload.description;
    }
    static makeError(error) {
        var _a, _b;
        const errorData = (_a = error === null || error === void 0 ? void 0 : error.response) === null || _a === void 0 ? void 0 : _a.data;
        if (this.isTravelTimeError(errorData)) {
            return new TravelTimeError(errorData);
        }
        return new TravelTimeError({
            description: 'Couldn\'t get data',
            http_status: ((_b = error === null || error === void 0 ? void 0 : error.response) === null || _b === void 0 ? void 0 : _b.status) || 500,
            error_code: 1,
            documentation_link: 'https://docs.traveltime.com/api/reference/error-codes',
            additional_info: {},
        });
    }
}
exports.TravelTimeError = TravelTimeError;
