// traveltime-api/target/client/rateLimiter.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RateLimiter = void 0;
const crypto_1 = __importDefault(require("crypto"));
class RateLimiter {
    constructor(rateLimitSettings) {
        this.setRateLimitSettings = (settings) => {
            this.rateLimitSettings = Object.assign(Object.assign({}, this.rateLimitSettings), settings);
        };
        this.requestQueue = [];
        this.completedQueue = new Set();
        this.isThrottleActive = false;
        this.isRequestInProgress = false;
        this.rateLimitSettings = Object.assign({ enabled: false, hitsPerMinute: 60 }, rateLimitSettings);
    }
    disableThrottle(hits) {
        if (!this.isThrottleActive)
            return;
        setTimeout(() => {
            this.isThrottleActive = false;
            if (this.requestQueue.length > 0)
                this.execute();
        }, ((60 * 1000) / this.rateLimitSettings.hitsPerMinute) * hits);
    }
    taskCleanUp(ids) {
        this.isRequestInProgress = false;
        if (this.requestQueue.length > 0)
            this.execute();
        setTimeout(() => {
            ids.forEach((id) => this.completedQueue.delete(id));
            this.execute();
        }, 1000 * 60);
    }
    execute() {
        return __awaiter(this, void 0, void 0, function* () {
            if (this.isRequestInProgress || this.isThrottleActive)
                return;
            const request = this.requestQueue.shift();
            if (!request)
                return;
            if (this.completedQueue.size + request.hits <= this.rateLimitSettings.hitsPerMinute) {
                this.isThrottleActive = true;
                this.isRequestInProgress = true;
                const uuids = [...Array(request.hits).keys()].map(() => crypto_1.default.randomUUID());
                uuids.forEach((id) => this.completedQueue.add(id));
                this.disableThrottle(uuids.length);
                yield request.task();
                this.taskCleanUp(uuids);
            }
            else {
                this.requestQueue.unshift(request);
            }
        });
    }
    addAndExecute(request, hits) {
        this.requestQueue.push({ task: request, hits });
        this.execute();
    }
    isEnabled() {
        return this.rateLimitSettings.enabled;
    }
}
exports.RateLimiter = RateLimiter;
